cpu/sam0_common/periph/freqm.c

```diff
@@ -0,0 +1,263 @@
+    bool freqm_is_busy = false;
+
+    if ((FREQM->STATUS.reg & FREQM_STATUS_BUSY_Msk) == FREQM_STATUS_BUSY_Msk) {
+        freqm_is_busy = true;
+    }
+
+    return freqm_is_busy;
+}
+
+int freqm_frequency_get(freqm_t idx, uint32_t *result)
+{
+    uint64_t result_tmp = 0U;
+    int overflow_condition = 0;
+    const struct freqm_config *config = &config_freqm[idx];
+
+    /* Deregister callback */
+    freqm_obj.callback = NULL;
+    freqm_obj.context = NULL;
+
+    _freqm_enable(config->ref_clk_cnt);
+
+    /* Check if measurement is already in progress */
+    if (!_freqm_is_busy()) {
+        /* Clear the Done Interrupt flag */
+        FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+        /* Start measurement */
+        FREQM->CTRLB.reg = (uint8_t)FREQM_CTRLB_START_Msk;
+    }
+
+    /* Block until measurement is done */
+    mutex_lock(&sync_lock);
+
+    /* Calculate measured frequency */
+    result_tmp = FREQM->VALUE.reg * (uint64_t)(sam0_gclk_freq(config->gclk_src));
+    result_tmp = result_tmp / config->ref_clk_cnt;
+
+    /* Return measured frequency */
+    *result = (uint32_t)result_tmp;
+
+    /* Read overflow status */
+    overflow_condition = ((int)FREQM->STATUS.reg & FREQM_STATUS_OVF_Msk);
+
+    /* Clear overflow status */
+    FREQM->STATUS.reg = (uint8_t)FREQM_STATUS_OVF_Msk;
+
+    _freqm_disable();
+
+    return overflow_condition;
+}
+
+void freqm_frequency_get_async(freqm_t idx, freqm_callback_t freqm_callback, void *context)
+{
+    const struct freqm_config *config = &config_freqm[idx];
+
+    /* Register callback function */
+    freqm_obj.callback = freqm_callback;
+    freqm_obj.context = context;
+    freqm_obj.idx = idx;
+
+    _freqm_enable(config->ref_clk_cnt);
+
+    /* Check if measurement is already in progress */
+    if (!_freqm_is_busy()) {
+        /* Clear the Done Interrupt flag */
+        FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+        /* Start measurement */
+        FREQM->CTRLB.reg = (uint8_t)FREQM_CTRLB_START_Msk;
+    }
+}
+
+void irq_freqm(void)
+{
+    /* Clear the Done Interrupt flag */
+    FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+    /* Invoked from non-blocking function */
+    if (freqm_obj.callback != NULL) {
+        if (_freqm_is_busy() == false) {
+            void *context_var;
+            const struct freqm_config *config = &config_freqm[freqm_obj.idx];
+            uint32_t result = 0;
+            uint64_t result_tmp = 0U;
+            int overflow_condition = 0;
+
+            /* Calculate measured frequency */
+            result_tmp = FREQM->VALUE.reg * (uint64_t)(sam0_gclk_freq(config->gclk_src));
+            result_tmp = result_tmp / config->ref_clk_cnt;
+            result = (uint32_t)result_tmp;
+
+            /* Read overflow status */
+            overflow_condition = ((int)FREQM->STATUS.reg & FREQM_STATUS_OVF_Msk);
+
+            /* Clear overflow status */
+            FREQM->STATUS.reg = (uint8_t)FREQM_STATUS_OVF_Msk;
```

Comment:
Instead of copying this code from `freqm_frequency_get()`, just move it to a common helper function that can be used by both.